src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.cc

```diff
@@ -0,0 +1,180 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <libbson-1.0/bson.h>
+
+#include <rapidjson/document.h>
+#include <string>
+
+#include "src/common/base/base.h"
+#include "src/common/base/utils.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/types.h"
+#include "src/stirling/utils/binary_decoder.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mongodb {
+
+ParseState ProcessOpMsg(BinaryDecoder* decoder, Frame* frame) {
+  PX_ASSIGN_OR(uint32_t flag_bits, decoder->ExtractLEInt<uint32_t>(), return ParseState::kInvalid);
+
+  // Find relevant flag bit information and ensure remaining bits are not set.
+  // Bits 0-15 are required and bits 16-31 are optional.
+  if (flag_bits & checksum_bitmask) {
+    frame->checksum_present = true;
+  }
+  if (flag_bits & more_to_come_bitmask) {
+    frame->more_to_come = true;
+  }
+  if (flag_bits & exhaust_allowed_bitmask) {
+    frame->exhaust_allowed = true;
+  }
+  if ((flag_bits >> 2) & required_unset_bitmask) {
+    return ParseState::kInvalid;
+  }
+
+  // Determine the number of checksum bytes in the buffer.
+  uint8_t checksum_bytes = (frame->checksum_present) ? 4 : 0;
+
+  // Get the section(s) data from the buffer.
+  while (decoder->BufSize() > checksum_bytes) {
+    mongodb::Section section;
+    PX_ASSIGN_OR(section.kind, decoder->ExtractLEInt<uint8_t>(), return ParseState::kInvalid);
+    // Length of the section still remaining in the buffer.
+    int32_t remaining_section_length = 0;
+
+    if (section.kind == 0) {
+      // Check the length but don't extract it since the later logic requires the buffer to retain
+      // it.
+      section.length = utils::LEndianBytesToInt<int32_t, 4>(decoder->Buf());
+      if (section.length < kSectionLengthSize) {
+        return ParseState::kInvalid;
+      }
+      remaining_section_length = section.length;
+
+    } else if (section.kind == 1) {
+      PX_ASSIGN_OR(section.length, decoder->ExtractLEInt<uint32_t>(), return ParseState::kInvalid);
+      if (section.length < kSectionLengthSize) {
+        return ParseState::kInvalid;
+      }
+
+      // Get the sequence identifier (command argument).
+      PX_ASSIGN_OR(std::string_view seq_identifier, decoder->ExtractStringUntil('\0'),
+                   return ParseState::kInvalid);
+      // Make sure the sequence identifier is a valid OP_MSG kind 1 command argument.
+      if (seq_identifier != "documents" && seq_identifier != "updates" &&
+          seq_identifier != "deletes") {
+        return ParseState::kInvalid;
+      }
+      remaining_section_length = section.length - kSectionLengthSize - seq_identifier.length() - 1;
+
+    } else {
+      return ParseState::kInvalid;
+    }
+
+    // Extract the document(s) from the section and convert it from type BSON to a JSON string.
+    auto parse_until = decoder->BufSize() - remaining_section_length;
```

Comment:
I find both of these `while` loops difficult to follow due to the variable naming. The first loop is for processing zero or more sections, while the second is for processing zero or more BSON objects/documents. Would we be able to try to align those high level concepts with the variable naming?

Happy to pair on this to discuss this.